### PR TITLE
Allow to authenticate with Vault through an IAM role assumed by Lambda function execution role

### PR DESCRIPTION
**Pull request by JinlianWang**
There are cases that we may want to be able to authenticate with Vault using another IAM role. For example, there are multiple Lambda functions (especially when a Lambda function is created for every HTTP method for the same RESTful resource, and/or historically an existing IAM role has been used for Vault authentication) that require access to the same set of secrets stored in Vault, instead of configuring access to every Lambda function using its execution role, an alternative is that every Lambda function assumes another IAM role which is used to authenticate with Vault. There are other cases that may make this authentication pattern more preferable, for example, a SaaS platform (say Snowflake or Saleforce) only require one IAM role to work with a tenant/client, instead of asking them for multiple IAM roles in order to retrieve secrets owned by each tenant/client. The same IAM role is assumed by different components on SaaS platform side in order to access secrets owned by each tenant/client.

### CODE REVIEW

**Comment by JinlianWang** [comment]
Example of Cloudwatch log (some of the sensitive infos are masked out using <info>): 

START RequestId: fbc175bb-7f57-454c-8713-f4b6a4b056c0 Version: $LATEST
2022-04-03T14:18:25.827Z [INFO]  vault-lambda-extension: Initialising
2022-04-03T14:18:25.827Z [DEBUG] vault-lambda-extension.vault-client: authenticating to Vault
2022-04-03T14:18:25.827Z [DEBUG] vault-lambda-extension.vault-client: **Trying to assume role with arn of arn:aws:iam::<account id>:role/Vault-Extension-Assumed-Role to authenticate with Vault**
2022-04-03T14:18:25.899Z [DEBUG] vault-lambda-extension.vault-client: **Assumed role successfully with token expiration time: 2022-04-03 15:18:25 +0000 UTC** 
2022-04-03T14:18:25.950Z [INFO]  vault-lambda-extension: Initialised
2022-04-03T14:18:25.950Z [INFO]  vault-lambda-extension: Starting HTTP proxy server
32081b6f-b427-49ce-ac77-c2577b2a9a4f2022-04-03T14:18:25.951Z [INFO]  vault-lambda-extension: Waiting for event...
EXTENSION	Name: vault-lambda-extension	State: Ready	Events: [INVOKE,SHUTDOWN]
2022-04-03T14:18:26.159Z [INFO]  vault-lambda-extension: Received event
2022-04-03T14:18:26.159Z [INFO]  vault-lambda-extension: Waiting for event...
2022-04-03T14:18:26.313Z [DEBUG] vault-lambda-extension.proxy: Proxying GET /v1/10e0a83d-fcaf-4bf6-8f6e-898b18254508/<secret path>
2022-04-03T14:18:26.375Z [DEBUG] vault-lambda-extension.proxy: Successfully proxied GET /v1/10e0a83d-fcaf-4bf6-8f6e-898b18254508/<secret path>
END RequestId: fbc175bb-7f57-454c-8713-f4b6a4b056c0
REPORT RequestId: fbc175bb-7f57-454c-8713-f4b6a4b056c0	Duration: 273.83 ms	Billed Duration: 274 ms	Memory Size: 128 MB	Max Memory Used: 88 MB	Init Duration: 449.69 ms

**Comment by steven-noorbergen** [comment]
We are running into the same issue as well, with multiple lambdas, and some lambdas not managed through terraform. This addition makes it easier for us to assign terraform-managed AWS roles, that through IAM policies are able to be assumed by lambdas that need to authenticate with vault.

**Comment by swenson** [comment]
Thanks for this. I'm working on testing it now, which is always a pain with STS + Lambda :)

**Comment by swenson** [comment]
Cool, this looks good after the conflict has been fixed and that error message is updated.

I was able to test this and it seemed to work well.

The modifications I made to the `quick-start` to test were (mostly):

```diff
iff --git a/quick-start/terraform/iam.tf b/quick-start/terraform/iam.tf
index 6a3881d..9b117b3 100644
--- a/quick-start/terraform/iam.tf
+++ b/quick-start/terraform/iam.tf
@@ -29,12 +29,24 @@ resource "aws_iam_role" "lambda" {
   assume_role_policy = data.aws_iam_policy_document.assume_role_lambda.json
 }
 
+resource "aws_iam_role" "extra_role" {
+  name               = "${var.environment_name}-extra-role"
+  assume_role_policy = data.aws_iam_policy_document.assume_role_lambda.json
+}
+
 resource "aws_iam_role_policy" "lambda" {
   name   = "${var.environment_name}-lambda-policy"
   role   = aws_iam_role.lambda.id
   policy = data.aws_iam_policy_document.lambda.json
 }
 
+resource "aws_iam_role_policy" "extra_role_policy" {
+  name   = "${var.environment_name}-extra-role-policy"
+  role   = aws_iam_role.extra_role.id
+  policy = data.aws_iam_policy_document.lambda.json
+}
+
+
 //--------------------------------------------------------------------
 // Data Sources
 
@@ -59,6 +71,11 @@ data "aws_iam_policy_document" "assume_role_lambda" {
       type        = "Service"
       identifiers = ["lambda.amazonaws.com"]
     }
+
+    principals {
+      type = "AWS"
+      identifiers = ["arn:aws:iam::${data.aws_caller_identity.current.account_id}:root"]
+    }
   }
 }
 
@@ -111,4 +128,10 @@ data "aws_iam_policy_document" "lambda" {
 
     resources = ["*"]
   }
+
+  statement {
+    effect  = "Allow"
+    actions = ["sts:AssumeRole"]
+    resources = ["*"]
+  }
 }
diff --git a/quick-start/terraform/lambda.tf b/quick-start/terraform/lambda.tf
index ebd0fde..69a578c 100644
--- a/quick-start/terraform/lambda.tf
+++ b/quick-start/terraform/lambda.tf
@@ -5,7 +5,7 @@ resource "aws_lambda_function" "function" {
   filename      = "../demo-function/demo-function.zip"
   handler       = "main"
   runtime       = "provided.al2"
-  layers        = ["arn:aws:lambda:${var.aws_region}:634166935893:layer:vault-lambda-extension:11"]
+  layers        = ["${aws_lambda_layer_version.vle.arn}"]
 
   environment {
     variables = {
@@ -15,7 +15,14 @@ resource "aws_lambda_function" "function" {
       VAULT_SECRET_PATH_DB = "database/creds/lambda-function",
       VAULT_SECRET_FILE_DB = "/tmp/vault_secret.json",
       VAULT_SECRET_PATH    = "secret/myapp/config",
+      VAULT_ASSUMED_ROLE_ARN = aws_iam_role.extra_role.arn,
       DATABASE_URL         = aws_db_instance.main.address
     }
   }
 }
+
+resource "aws_lambda_layer_version" "vle" {
+  filename = "vle.zip"
+  layer_name = "vault-lambda-extension"
+  compatible_architectures = ["x86_64"]
+}
\ No newline at end of file
diff --git a/quick-start/terraform/rds.tf b/quick-start/terraform/rds.tf
index 78cd1fa..ca6a748 100644
--- a/quick-start/terraform/rds.tf
+++ b/quick-start/terraform/rds.tf
@@ -7,7 +7,7 @@ resource "aws_db_instance" "main" {
   allocated_storage      = 20
   storage_type           = "gp2"
   engine                 = "postgres"
-  engine_version         = "12.3"
+  engine_version         = "12.9"
   instance_class         = var.db_instance_type
   name                   = "lambdadb"
   username               = "vaultadmin"
diff --git a/quick-start/terraform/templates/userdata-vault-server.tpl b/quick-start/terraform/templates/userdata-vault-server.tpl
index 05cf4b7..743f090 100755
--- a/quick-start/terraform/templates/userdata-vault-server.tpl
+++ b/quick-start/terraform/templates/userdata-vault-server.tpl
@@ -182,7 +182,7 @@ vault auth enable aws
 vault write -force auth/aws/config/client
 vault write auth/aws/role/${tpl_role_name} \
   auth_type=iam \
-  bound_iam_principal_arn="arn:aws:iam::${tpl_account_id}:role/${tpl_role_name}" \
+  bound_iam_principal_arn="${tpl_bound_role}" \
   policies=lambda-function \
   ttl=24h
 
diff --git a/quick-start/terraform/vault-server.tf b/quick-start/terraform/vault-server.tf
index 70dad45..de63b33 100644
--- a/quick-start/terraform/vault-server.tf
+++ b/quick-start/terraform/vault-server.tf
@@ -22,6 +22,7 @@ resource "aws_instance" "vault-server" {
       tpl_aws_region         = var.aws_region
       tpl_account_id         = data.aws_caller_identity.current.account_id
       tpl_role_name          = aws_iam_role.lambda.name
+      tpl_bound_role         = aws_iam_role.extra_role.arn
       tpl_rds_endpoint       = aws_db_instance.main.endpoint
       tpl_rds_admin_password = random_password.password.result
   })
diff --git a/quick-start/terraform/versions.tf b/quick-start/terraform/versions.tf
index dd5efa0..e011bc0 100644
--- a/quick-start/terraform/versions.tf
+++ b/quick-start/terraform/versions.tf
@@ -3,7 +3,7 @@ terraform {
   required_version = ">= 0.12"
   required_providers {
     aws = {
-      version = "~> 3.39.0"
-    } 
+      version = "~> 4.24.0"
+    }
   }
 }
```

**Comment by swenson** [comment]
@JinlianWang are you able to fix the merge conflicts, or would you like me to?

**Comment by JinlianWang** [comment]
@swenson sorry I was not knowing that you were expecting me to resolve the conflict. I am inclined to that you do whatever it takes to merge it, including throwing away all my code or any part of it :-)

**Comment by swenson** [comment]
@JinlianWang I think this is good overall! If you would resolve the conflict in the README I can merge it. I'll likely come in and add the test I wrote after.

File: internal/vault/client.go [hunk 1]
```diff
@@ -93,8 +97,49 @@ func (c *Client) Token(ctx context.Context) (string, error) {
...
+
+		sessionName := "vault_auth"
+
+		result, err := c.stsSvc.AssumeRole(&sts.AssumeRoleInput{
+			RoleArn:         &roleToAssumeArn,
+			RoleSessionName: &sessionName,
+		})
+
+		if err != nil {
+			return fmt.Errorf("failed to assume role with arn of %s", roleToAssumeArn)
```
[the remark this answers is not in the document]
Comment (JinlianWang) [reply internal/vault/client.go]: Good point. Made the change.

**Comment by JinlianWang** [comment]
@swenson Done. Just want to make sure that a change will be made to update the general [usage documentation](https://www.vaultproject.io/docs/platform/aws/lambda-extension): 

`VAULT_ASSUMED_ROLE_ARN` | Valid ARN of an IAM role that can be assumed by the execution role assigned to your Lambda function. | No | `arn:aws:iam::123456789012:role/xaccounts3access`

**Comment by swenson** [comment]
Awesome. Thanks! I'll update the usage documentation and do some testing.

**Review by swenson — APPROVED** [review]

**Comment by JinlianWang** [comment]
@steven-noorbergen just want to make sure that you are aware of the merge.

**Pull request merged**